8287087: C2: perform SLP reduction analysis on-demand

robcasloz opened the pull request:
Reduction analysis finds cycles of reduction operations within loops. The result of this analysis is used by SLP auto-vectorization (to vectorize reductions if deemed profitable) and by x64 instruction matching (to select specialized scalar floating-point `Math.min()/max()` implementations). Currently, reduction analysis is applied early (before loop unrolling), and the result is propagated through loop unrolling by marking nodes and loops with special reduction flags. Applying reduction analysis early is efficient, but propagating the results correctly through loop unrolling and arbitrary graph transformations is challenging and often leads to inconsistent node-loop reduction flag states, some of which have led to actual miscompilations in the past (see [JDK-8261147](https://bugs.openjdk.org/browse/JDK-8261147) and [JDK-8279622](https://bugs.openjdk.org/browse/JDK-8279622)).

This changeset postpones reduction analysis to the point where its results are actually used. To do so, it generalizes the analysis to find reduction cycles on unrolled loops:

![reduction-before-after-unrolling](https://user-images.githubusercontent.com/8792647/226725587-b7d68509-3717-4bbe-8d54-f9a105853fda.png)

The generalized analysis precludes the need to maintain and propagate node and loop reduction flags through arbitrary IR transformations, reducing the risk of miscompilations due to invalidation of the analysis results. The generalization is slightly more costly than the current analysis, but still negligible in micro- and general benchmarks.

## Performance Benefits

 As a side benefit, the proposed generalization is able to find more reductions, increasing the scope of auto-vectorization and the performance of x64 floating-point `Math.min()/max()` in multiple scenarios.

### Increased Auto-Vectorization Scope

There are two main scenarios in which the proposed changeset enables further auto-vectorization:

#### Reductions Using Global Accumulators

```
public class Foo {
  int acc = 0;
  (..)
  void reduce(int[] array) {
    for (int i = 0; i < array.length; i++) {
      acc += array[i];
    }
  }
}
```
Initially, such reductions are wrapped by load and store nodes, which defeats the current reduction analysis. However, after unrolling and other optimizations are applied, the reduction becomes recognizable by the proposed analysis:

![global-reduction-before-after-unrolling](https://user-images.githubusercontent.com/8792647/226745351-33494e40-7c07-4a8b-8bf6-d3a96e84b1c2.png)

#### Reductions of partially unrolled loops

```
    (..)
    for (int i = 0; i < array.length / 2; i++) {
      acc += array[2*i];
      acc += array[2*i + 1];
    }
    (..)
```

These reductions are manually unrolled from the beginning, so the current reduction analysis fails to find them, while the proposed analysis is able to detect them as if they were unrolled automatically.

### Increased Performance of x64 Floating-Point `Math.min()/max()`

Besides the above scenarios, the proposed generalization allows the x64 matcher to select specialized floating-point `Math.min()/max()` implementations for reductions in non-counted and outer loops (see the new micro-benchmarks in `FpMinMaxIntrinsics.java` for more details).

## Implementation details

The generalized reduction analysis finds reductions in a loop by looking for chains of reduction operators of the same node type starting and finishing on each phi node in the loop. To avoid a combinatorial explosion, the analysis assumes that all nodes in a chain are connected via the same edge index, which is realistic because chains usually consist of identical nodes cloned by loop unrolling. This assumption allows the analysis to test only two paths for each examined phi node. A failure of this assumption (e.g. as illustrated in test case `testReductionOnPartiallyUnrolledLoopWithSwappedInputs` from `TestGeneralizedReductions.java`) results in mising vectorization but does not affect correctness. Note that the same-index assumption can only fail in cases where current auto-vectorization would also fail to vectorize (manually unrolled loops).

The changeset implements a more relaxed version of the reduction analysis for x64 matching, suitable for queries on single nodes. This analysis is run only in the presence of `[Min|Max][F|D]` nodes.

## Alternative approaches

A complication results from edge swapping in the nodes cloned by loop unrolling (see [here](https://github.com/openjdk/jdk/blob/bbde2158d1d11be909292d0c8625211e6cf5359e/src/hotspot/share/opto/addnode.cpp#L123) and [here](https://github.com/openjdk/jdk/blob/bbde2158d1d11be909292d0c8625211e6cf5359e/src/hotspot/share/opto/mulnode.cpp#L113)), which can lead to reduction chains connected via different input indices. This is addressed by tracking whether nodes have swapped edges and adjusting the explored input indices in the reduction analysis accordingly. An alternative (proposed by @eme64 and @jatin-bhateja ) is to replace this changeset's linear chain finding approach with some form of general path-finding algorithm. This alternative would preclude the need for tracking edge swapping at a potentially higher computational cost. The following table summarizes the pros and cons of the current mainline approach, this changeset, and the proposed alternative:

| approach | correctness | efficiency | effectiveness | conceptual complexity |
| -------- | ----------- | ---------- | ------------- | --------------------- |
| mainline (current) | hard to establish due to need of maintaining reduction flags through arbitrary graph transformations (has led to miscompilations, see JDK-8261147 and JDK-8279622) | high | low (misses substantial reduction vectorization opportunities) | high (requires maintaining non-local reduction node state) |
| this changeset | easy to establish since client transformations operate on the same graph that is analyzed | medium (limited search for chains of nodes) | high (finds all reduction cycles except for partially unrolled loops with manually-swapped inputs) | medium (requires maintaining local swapped-edge node state) |
| general search | easy to establish (same as above) | low (general search), particularly for x64 matching where the analysis runs once for every node in a chain | high (similar to above but also covering manually-swapped inputs) | low (no node state required, use of well-known graph search algorithms) |

Since the efficiency-conceptual complexity trade-off between this changeset and the general search approach is not obvious, I propose to integrate this changeset (which strikes a balance between the two) and investigate the latter one in a follow-up RFE.

## Testing

### Functionality

- tier1-5 (linux-x64, linux-aarch64, windows-x64, macosx-x64, and macosx-aarch64).
- fuzzing (12 h. on linux-x64 and linux-aarch64).

##### TestGeneralizedReductions.java

Tests the new scenarios in which vectorization occurs. These tests are restricted to 64-bits platforms, since I do not have access to 32-bits ones. `testReductionOnPartiallyUnrolledLoop` has been observed to fail on [linux-x86](https://github.com/robcasloz/jdk/actions/runs/4478959520/jobs/7873827856#logs) due to missing vectorization. If anyone wants to have a look and derive the necessary IR test framework preconditions for the test to pass on linux-x86, I am happy to lift the 64-bits restriction.

##### TestFpMinMaxReductions.java

Tests the matching of floating-point max/min implementations in x64.

##### TestSuperwordFailsUnrolling.java

This test file is updated to ensure auto-vectorization is never triggered, because this changeset would otherwise enable it and defeat the purpose of the test.

### Performance

#### General Benchmarks

The changeset does not cause any performance regression on the DaCapo, SPECjvm 2008, and SPECjbb2015 benchmark suites for linux-x64 and linux-aarch64.

#### Micro-benchmarks

The changeset extends two existing files with additional micro-benchmarks that show the benefit of the generalized reduction analysis ([full results](https://github.com/openjdk/jdk/files/11039207/microbenchmark-results.ods)).


##### VectorReduction.java

These micro-benchmarks are first adjusted to actually vectorize in the mainline approach, since they suffered from the global-accumulator limitation. Two micro-benchmarks are added to exercise vectorization in the presence of global accumulators and partially unrolled loops. Running `VectorReduction.java` on an x64 (Cascade Lake) machine confirms the expectations: compared to mainline (with the adjustment mentioned above), this changeset yields similar performance results except for `andRedIOnGlobalAccumulator` and `andRedIPartiallyUnrolled`, where the changeset improves performance by 2.4x in both cases.

##### MaxIntrinsics.java

This file is extended with four new micro-benchmarks. Running it on the same machine as above shows that the changeset does not affect the performance of the existing micro-benchmarks, and improves moderately to substantially the performance of the new ones (because it allows the x64 matcher to select a floating-point `Math.min()` implementation that is specialized for reduction min operations):

| micro-benchmark             | speedup compared to mainline |
| ---                         | ---                          |
| `fMinReduceInOuterLoop`       | 1.1x                         |
| `fMinReduceNonCounted`        | 2.3x                         |
| `fMinReduceGlobalAccumulator` | 2.4x                         |
| `fMinReducePartiallyUnrolled` | 3.9x                         |

## Acknowledgments

Thanks to @danielogh for making it possible to test this improvement with confidence ([JDK-8294715](https://bugs.openjdk.org/browse/JDK-8294715)) and to @TobiHartmann, @chhagedorn, @vnkozlov and @eme64 for discussions and useful feedback.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287087](https://bugs.openjdk.org/browse/JDK-8287087): C2: perform SLP reduction analysis on-demand


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - Committer)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13120/head:pull/13120` \
`$ git checkout pull/13120`

Update a local copy of the PR: \
`$ git checkout pull/13120` \
`$ git pull https://git.openjdk.org/jdk.git pull/13120/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13120`

View PR using the GUI difftool: \
`$ git pr show -t 13120`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13120.diff">https://git.openjdk.org/jdk/pull/13120.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13120#issuecomment-1479369133)

eme64 commented on test/hotspot/jtreg/compiler/loopopts/superword/TestGeneralizedReductions.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Test
+    @IR(applyIfCPUFeatureAnd = {"sse4.1", "true", "avx2", "true"},
```
Does `avx2` not imply `sse4.1`?

eme64 commented on src/hotspot/share/opto/node.hpp:
```diff
@@ -578,6 +578,12 @@ class Node {
+      remove_flag(Node::Flag_has_swapped_edges);
+    } else {
+      add_flag(Node::Flag_has_swapped_edges);
+    }
```
This is probably the riskiest part of the implementation. If anyone decides to swap edges without using `swap_edges`, we will not detect the reduction. We have to trust our tests to catch that.

robcasloz replied:
That's right, in that case we would miss the reduction. On the other hand, the assumption that edges are ordered in the current form is deeply ingrained in C2, and if this was altered, missing some reduction vectorizations would be the least of our problems. And, as you mention, thanks to the introduction of IR checks in the reduction tests we would detect the missing vectorization anyway.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -421,6 +426,147 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+void SuperWord::mark_reductions(IdealLoopTree* loop) {
+
+  _loop_reductions.clear();
+  const CountedLoopNode* loop_head = loop->_head->as_CountedLoop();
```
Can you not use the SuperWord member function `lp()`?

robcasloz replied:
Good suggestion, will do.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -421,6 +426,147 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+    for (int i = 0; i < path_nodes; i++) {
+      for (DUIterator_Fast jmax, j = current->fast_outs(jmax); j < jmax; j++) {
+        Node* u = current->fast_out(j);
+        if (!loop->is_member(_phase->get_loop(_phase->ctrl_or_self(u)))) {
```
same

robcasloz replied:
Will do.

robcasloz commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -421,6 +426,147 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+          [&](const Node* n) {
+            Node* ctrl = _phase->get_ctrl(n);
+            return (n->Opcode() == first->Opcode() && ctrl != nullptr &&
+                    loop->is_member(_phase->get_loop(ctrl)));
```
Will do.

robcasloz commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -421,6 +426,147 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+  const Node* trip_phi = loop_head->phi();
+
+  // Iterate through all phi nodes associated to the loop and search for
+  // reduction cycles of at most LoopMaxUnroll nodes.
```
Thanks for the suggestion! I prefer to keep `LoopMaxUnroll` as a bound, to guard against pathologically large basic blocks, even if it implies missing some hand-unrolled loops.

robcasloz replied:
Good point. I will test using the loop body size as a bound for superword reduction analysis, but leave the tighter `LoopMaxUnroll` bound for selecting floating-point min/max implementations, which is more time sensitive.

robcasloz commented:
I just pushed a new version addressing @eme64's comments and suggestions, please review.

robcasloz commented:
> Thanks for the updates @robcasloz .
> Looks good to me now.

Thanks for reviewing, Emanuel!

jatin-bhateja commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -411,6 +416,139 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+    }
+    // Test that reduction nodes do not have any users in the loop besides their
+    // reduction cycle predecessors.
+    const Node* current = first;
+    const Node* pred = phi; // current's predecessor in the reduction cycle.
+    bool used_in_loop = false;
+    for (int i = 0; i < path_nodes; i++) {
+      for (DUIterator_Fast jmax, j = current->fast_outs(jmax); j < jmax; j++) {
+        Node* u = current->fast_out(j);
+        if (!in_bb(u)) {
+          continue;
+        }
+        if (u == pred) {
+          continue;
+        }
+        used_in_loop = true;
+        break;
+      }
+      if (used_in_loop) {
+        break;
+      }
+      pred = current;
+      current = original_input(current, reduction_input);
+    }
```
If we bookkeep the nodes in the reduction chain path during initial backward traversal we may simplify this checking and also another call to _original_input_ while populating _loop_reductions set on [#L547 ](https://github.com/openjdk/jdk/pull/13120/files#diff-8f29dd005a0f949d108687dabb7379c73dfd85cd782da453509dc9b6cb8c9f81R547)

robcasloz replied:
Thanks, will consider this together with your earlier comments.

robcasloz replied:
I tried out your suggestion but unfortunately, the bookkeeping code (marking/storing candidate nodes and their predecessors in the tentative reduction chain) became more complex than the simplifications it enabled.

robcasloz replied:
Thanks for the patch, it is very similar to what I had tried before (see my comment above), except your patch rejects reduction chains with external users, whereas this PR allows reduction chain nodes to be used as long as the user is not within the loop. This follows the original logic more closely (although I have not found a case where the distinction matters yet):

https://github.com/openjdk/jdk/blob/a4a5385831b58e66fe3f34cef618643f9be68c9e/src/hotspot/share/opto/loopTransform.cpp#L2539-L2549)

To ease comparison, I adapted your patch to perform the same test as this PR. Here is the result:

https://github.com/robcasloz/jdk/compare/JDK-8287087...robcasloz:jdk:JDK-8287087-single-original-input-call

In my opinion, even if the patch gets rid of the last reduction path traversal, the result is not necessarily more readable or efficient, due to the cognitive and computational overhead of introducing an auxiliary `GrowableArray` data structure.

jatin-bhateja commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -411,6 +416,139 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+      // Test whether there is a reduction path in the basic block from 'first'
+      // to the phi node following edge index 'input'.
+      PathEnd path =
+        find_in_path(
```
One way to add fault-tolerance to find_in_path could be to follow strict DFS semantics where an alternate path is taken if node's predicates are not satisfied, currently we are starting all over again from the first node of chain with a different reduction_input which prevents inferring reduction chain even though all the nodes in the chain are commutative isomorphic operations.

robcasloz replied:
Thanks for the observation and the suggestion! This changeset proposes swapped edge tracking for efficiency and better worst-case behavior of the analysis, but as you observe this is done at the expense of robustness. I will investigate whether the swapped edge tracking approach can be extended to deal with the `MinI`/`MaxI` transformations that you mention. If not, I will re-consider using a generic search approach like you and @eme64 suggest.

robcasloz replied:
Hi again @jatin-bhateja,

I have studied now [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673) in more detail, and my conclusion is that it is not a duplicate but actually orthogonal to this changeset. Even a perfect reduction analysis alone would not re-enable the missing vectorization, because the canonicalization transformations done by `MaxI/MinINode::Ideal()` inhibit SuperWord analysis at a later stage.

In light of this, I propose to re-open [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673), and address it by handling all four combinations of two-level inputs in `MaxI/MinINode::Ideal()` instead of canonicalizing `MaxI/MinI` chains. This solution is, in my opinion, more straightforward (and not necessarily more expensive). The main reason is that it separates concerns, making it possible to reason about input swapping, optimization of `MaxI/MinI` nodes, reduction analysis, and auto-vectorization separately. I have [a WIP, prototype implementation](https://github.com/openjdk/jdk/compare/master...robcasloz:jdk:JDK-8302673) which seems to work fine for all of the discussed reduction analysis strategies. @jatin-bhateja, if you want I can take over JDK-8302673 and submit it for review once I have polished it.

Given that JDK-8302673 is orthogonal to this RFE, and that a solution to JDK-8302673 is available for which this RFE detects MaxI/MinI reductions correctly, I suggest to move on with this RFE as-is, and file a follow-up RFE to investigate generic search approaches.

robcasloz commented:
Moving back to draft mode for further investigation and conflict resolution.

eme64 commented:
I filed this RFE, it is related to this work here: [JDK-8305707](https://bugs.openjdk.org/browse/JDK-8305707) "SuperWord should vectorize reverse-order reduction loops"

robcasloz commented:
I have resolved conflicts caused by the integration of [JDK-8304042](https://bugs.openjdk.org/browse/JDK-8304042); added minimal, debug-only code for emitting `Node::Flag_has_swapped_edges` for IGV nodes; and addressed @jatin-bhateja's comments (including analyzing the interaction with [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673) and determining it is orthogonal to this RFE). Please review.

robcasloz commented:
Hi @jatin-bhateja, I have written a qualitative comparison between this PR and the generic search approach proposed by @eme64 and you (see **Alternative approaches** section in the updated PR description). I hope the comparison clarifies and motivates the plan outlined in https://github.com/openjdk/jdk/pull/13120#discussion_r1166778814. Please let me know whether you agree with that plan so that we can move forward with this RFE, [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673), and also [JDK-8302652](https://bugs.openjdk.org/browse/JDK-8302652).

jatin-bhateja commented:
> Hi @jatin-bhateja, I have written a qualitative comparison between this PR and the generic search approach proposed by @eme64 and you (see **Alternative approaches** section in the updated PR description). I hope the comparison clarifies and motivates the plan outlined in [#13120 (comment)](https://github.com/openjdk/jdk/pull/13120#discussion_r1166778814). Please let me know whether you agree with that plan so that we can move forward with this RFE, [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673), and also [JDK-8302652](https://bugs.openjdk.org/browse/JDK-8302652).

Hi @robcasloz , Problem occurs due to Min/Max canonicalizing transformations which results into creation of new nodes but does not propagate the _has_swapped_edges_ flags. A forward traversal starting from output of phi node can avoid edge swapping related issues and can give up discovering a path if any node feeds more than one users, want to stress that even if _mark_reductions_ detects a set of nodes as part of reduction chain SLP may still not vectorize it e.g. an AddI reduction chain with different constant inputs. 

Your approach looks good to me, path finding is strict and flows same edge for path discovery and fixes several missed reduction scenarios.

robcasloz commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -411,6 +416,139 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+      continue;
+    }
+    // Test that reduction nodes do not have any users in the loop besides their
+    // reduction cycle predecessors.
```
Good catch, thanks! My confusion stemmed from thinking about the traversal order rather than the direction of the edges in the Ideal graph. I have updated the comment and the variable name.

jatin-bhateja reviewed (approved):
Hi @robcasloz , Apart from some concern on path detection traversals which are not blocking issues patch looks good to me. 
Best Regards,
Jatin

robcasloz commented:
> Still looks good.

Thanks for looking at this again, Emanuel!

TobiHartmann reviewed (approved):
Great, thorough analysis. The fix looks good to me.

robcasloz commented:
> Great, thorough analysis. The fix looks good to me.

Thanks for reviewing, Tobias!

robcasloz commented:
/integrate

robcasloz commented:
> Since the efficiency-conceptual complexity trade-off between this changeset and the general search approach is not obvious, I propose to integrate this changeset (which strikes a balance between the two) and investigate the latter one in a follow-up RFE.

Filed now: [JDK-8306989](https://bugs.openjdk.org/browse/JDK-8306989).
